plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py

```diff
@@ -1110,62 +882,43 @@ def run(
-        self.grid_ind = np.where(self.theta < self.max_theta_hit_grid)[0]
-        self.nparticles_grid = len(self.grid_ind)
-        self.fract_tracked = self.nparticles_grid / self.nparticles
-
-        # Create flags for tracking when particles during the simulation
-        # on_grid -> zero if the particle is off grid, 1
-        # shape [nparticles, ngrids]
-        self.on_grid = np.zeros([self.nparticles_grid, self.num_grids])
-        # Entered grid -> non-zero if particle EVER entered a grid
-        self.entered_grid = np.zeros([self.nparticles_grid])
+        # Determine which particles should be removed
+        # These particles will not be pushed through the fields
+        theta_mask = self.theta >= self.max_theta_hit_grid
+        self.x = self._coast_to_plane(
+            self.detector, self.det_hdir, self.det_vdir, mask=theta_mask
+        )
+        self._stop_particles(theta_mask)
+        self.fract_tracked = self.nparticles_tracked / self.nparticles
 
         # Generate a null distribution of points (the result in the absence of
         # any fields) for statistical comparison
         self.x0 = self._coast_to_plane(self.detector, self.det_hdir, self.det_vdir)
 
-        # Advance the particles to the near the start of the grid
+        # Advance the tracked particles to the near the start of the grid
         self._coast_to_grid()
+        self.coasted_particles = np.copy(self.x)
 
-        # Initialize a "progress bar" (really more of a meter)
-        # Setting sys.stdout lets this play nicely with regular print()
-        pbar = tqdm(
-            initial=0,
-            total=self.nparticles_grid + 1,
-            disable=not self.verbose,
-            desc="Particles on grid",
-            unit="particles",
-            bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",
-            file=sys.stdout,
-        )
-
-        # Push the particles until the stop condition is satisfied
-        # (no more particles on the simulation grid)
-        while not self._stop_condition():
-            n_on_grid = np.sum(self.on_any_grid)
-            pbar.n = n_on_grid
-            pbar.last_print_n = n_on_grid
-            pbar.update()
+        super().run()
 
-            self._push()
-        pbar.close()
+        if self.num_entered < 0.1 * self.nparticles:
+            warnings.warn(
+                f"Only {100 * self.num_entered / self.nparticles:.2f}% of "
+                "particles entered the field grid: consider "
+                "decreasing the max_theta to increase this "
+                "number.",
+                RuntimeWarning,
+            )
 
         # Remove particles that will never reach the detector
         self._remove_deflected_particles()
 
         # Advance the particles to the image plane
         self._coast_to_plane(self.detector, self.det_hdir, self.det_vdir, x=self.x)
 
-        # Log a summary of the run
+        self.save_routine.post_push_hook(force_save=True)
```

Comment:
I still think here it would be OK to just call `save` explicitly? What is the post_push_hook method doing for us now?